FEATURE (WIP): add max-width and center email notifications

scossar opened the pull request:
Work in progress:

- adds a layout that sets a max-width of 600px (and a fixed width of 600px for Outlook) see: http://webdesign.tutsplus.com/tutorials/creating-a-future-proof-responsive-email-without-media-queries--cms-23919

- sets a max-width of 600px on images (fixed width of 600px for Outlook)

The css in the layout is written inline for now. It could be added to `Email::Styles`.

This may work with a more minimal layout that doesn't add a `head` section or doctype. I have read that there are issues with the iphone not respecting padding on content when there is no `<meta name="viewport" content="width=device-width, initial-scale=1.0">` tag. I haven't tested that though.

Using `Margin` instead of `margin` is a hack to keep it from being stripped by Outlook.

Feel free to use any of this.

![screenshot 2016-01-20 13 57 04](https://cloud.githubusercontent.com/assets/2975917/12465072/2ab0ed78-bf81-11e5-82e7-30bb540d0346.png)

![screenshot 2016-01-20 13 57 14](https://cloud.githubusercontent.com/assets/2975917/12465090/3d5579a8-bf81-11e5-961f-c0fb9ba7918d.png)

![screenshot 2016-01-20 13 57 39](https://cloud.githubusercontent.com/assets/2975917/12465098/4a7422f6-bf81-11e5-9ede-de5ac71080b3.png)

![screenshot 2016-01-20 13 58 31](https://cloud.githubusercontent.com/assets/2975917/12465109/52b22f76-bf81-11e5-88ee-91509e14beab.png)

coding-horror commented:
do you feel this is safe to merge?

scossar commented:
No. Someone needs to look at how it's adding the layout in
`user_notifications` (it's doing it in two places) and at the change in
`styles.rb` from using `@fragment = Nokogiri::HTML.fragment(@html)` to
`@doc= Nokigiri::HTML(@html)`.

On Wed, Jan 20, 2016 at 3:58 PM, Jeff Atwood <notifications@github.com>
wrote:

> do you feel this is safe to merge?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/discourse/discourse/pull/3969#issuecomment-173406891>.
>

coding-horror commented:
ok I nominate @ZogStriP as he's on mail as a primary feature for a few more weeks..

scossar commented:
If you just want to use the ideas in there that's fine by me. The layout seems good. It's what's recommended in the tutorial I reference in the first comment on here, and is very close to what they teach in the CodeSchool email course. Setting the max-width on the images for Outlook (and Yahoo) was a bit of trial end error. It's set in `Styles::format_basic`
```
...
      big_images.each do |img|
        add_styles(img, 'width: 100%; height: auto; max-width: 600px;') if img['style'] !~ /max-width/
        add_attributes(img, width: 600)
      end
```

The logic is that Outlook gets its width from the `width="600"` attribute, everything else gets the width from the inline styles. If `max-width` isn't set images can get huge in Yahoo.

ZogStriP commented:
I like it, but you know what's missing? Some tests ;) Can you add some?

scossar commented:
Sure, I'd like to. It might be a couple of days before I can get it done.

On Thu, Jan 21, 2016 at 7:21 AM, Régis Hanol <notifications@github.com>
wrote:

> I like it, but you know what's missing? Some tests ;) Can you add some?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/discourse/discourse/pull/3969#issuecomment-173603931>.
>

scossar commented:
@ZogStriP I've added some tests.

I reworked `styles_spec.rb` so that it works with a Nokogiri document instead of a fragment. Under `context "format notifications"` I'm testing that I haven't broken `Styles#style`.

I had a hard time finding a way to test that that correct layout is being used for email notifications. I ended up checking for the meta tag from the layout. Maybe there's a better way. 

Any suggestions for improvement are welcome.

I have access to Litmus email testing for the next couple of weeks - possibly longer if I get some work using it. I did a quick test of a post. You can see the results here: https://litmus.com/checklist/public/b5e45d3 
I'm happy to tweak the styles if there's anything you want changed - (font-size and line height could be set for desktop clients.)

If you want to add a basic layout to notifications that don't have an html_part, (so that their width is constrained) that can be done in `renderer.rb` by wrapping `PrettyText.cook(text)` in a table.

coding-horror commented:
if @ZogStriP  is OK with it we can merge

ZogStriP commented:
Thanks :+1:
